doc/antora/modules/howto/pages/data-usage-reporting/index.adoc

```diff
@@ -0,0 +1,222 @@
+existing sessions will of course result in an interruption of connectivity for
+users.
+
+
+== Solution
+
+FreeRADIUS is distributed with a schema extension which introduces a
+`data_usage_by_period` table containing data usage keyed by time-period and
+user, and provides the means of “running an update process” to correctly
+populate this table, either by calling a stored procedure or running a seperate
+script:
+
+    [raddb]/mods-config/sql/main/[database]/process-radacct.sql
+
+Rather than burdening the server with populating per-user, time-period data in
+real time, the standard `radacct` table is periodically processed to extract the
+session data that was updated since the previous update was ran.
+
+Running an update closes the current “accounting period“ allocating data
+consumed by all sessions that “reported in” during that period (via
+interim-updates or stops) to the period’s users. The update process
+additionally creates an open-ended, next accounting period to capture future
+data usage. It is careful to avoid duplicate counting of data used in sessions
+that span multiple accounting periods by carrying forward into this next period
+a negated amount for the data that was already allocated to any earlier
+periods.
+
+The update process appends a number of rows to the `data_usage_by_period` table
+that is equal to the number of users whose sessions were active within the
+current accounting period and therefore makes efficient use of storage.
+
+NOTE: The update process does not have to be called at regular intervals but in
+production is likely to be invoked either as a cronjob or using the database
+server’s event scheduler. It can be safely manually invoked (for debugging or
+ad hoc reporting purposes) as often as necessary outside of the normal schedule
+without impacting normal usage reporting.
+
+For reporting purposes, the adhoc periods within the `data_usage_by_period`
+table can be efficiently aggregated to provide data bucketed into the required
+reporting intervals.
+
+
+== Example setup
+
+Megacorp stores their RADIUS accounting data in a MySQL database using the
+standard FreeRADIUS schema. They want to provide reports of monthly data usage
+for given users.
+
+Firstly, implement the extended schema to create the `data_usage_by_period`
+table to hold the data usage amounts and the `fr_new_data_usage_period` stored
+procedure that updates this table:
+
+    mysql radius < /etc/raddb/mods-config/sql/main/mysql/process-radacct.sql
+
+NOTE: The above command assumes that passwordless login has been configured via
+the user's `~/.my.cnf` file, or otherwise. This command and subsequent commands
+should be adjusted accordingly where this is not the case.
+
+Call the update process manually to create the initial accounting period for
+the historical data usage and start of the next period for future use:
+
+    $ echo “CALL fr_new_data_usage_period()” | mysql radius
```

Comment:
Needs a code block (shell), fix smart quotes, remove dollar (means no one can paste the example into a terminal)